### PR TITLE
Add in basic GPU/CPU bridge operation [databricks]

### PR DESCRIPTION
**Pull request by revans2**
This is step 3 in splitting https://github.com/NVIDIA/spark-rapids/pull/13368 into smaller pieces

### Description

This adds in basic GPU/CPU bridge functionality, but it is off by default because the performance would not be good without the thread pool and optimizer.

### Checklists

- [ ] This PR has added documentation for new or modified features or behaviors.
- [X] This PR has added new tests or modified existing tests to cover new code paths.
      (Please explain in the PR description how the new code paths are tested, such as names of the new/existing tests that cover them.)
- [ ] Performance testing has been performed and its results are added in the PR description. Or, an issue has been filed with a link in the PR description.

The performance is expected to be bad so it is off by default an not tested. I did add some basic tests to verify that the code works.

### CODE REVIEW

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
CI failed because of https://github.com/NVIDIA/spark-rapids/issues/14009 but our CI currently has not way to turn off the databricks tests when you touched something even remotely related to databricks.

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
I upmerged to make sure that I had the latest fixes for what caused CI to fail

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/TypeConverter.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.io.Serializable;
+
+import org.apache.spark.sql.catalyst.expressions.SpecializedGetters;
+
+public abstract class TypeConverter implements Serializable {
+  /**
+   * Append row value to the column builder and return the number of data bytes written
+   */
+  public abstract double append(SpecializedGetters row, int column, RapidsHostColumnBuilder builder);
```
Comment (abellina) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/TypeConverter.java:28]: it would be nice to add a comment about the return type here (double). Why can't the type be for a whole number?

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/TypeConverter.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
```
Comment (abellina) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/TypeConverter.java]: copyrights need updating.

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/TypeConverter.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  public void appendNoRet(SpecializedGetters row, int column, RapidsHostColumnBuilder builder) {
+    append(row, column, builder);
+  }
+
+  /**
+   * This is here for structs.  When you append a null to a struct the size is not known
+   * ahead of time.  Also because structs push nulls down to the children this size should
+   * assume a validity even if the schema says it cannot be null.
+   */
+  public abstract double getNullSize();
```
Comment (abellina) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/TypeConverter.java:44]: same question, why double?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  /**
+   * Evaluate the expression sequentially (single-threaded).
+   */
+  private def evaluateSequentially(batch: ColumnarBatch): GpuColumnVector = {
+    val numRows = batch.numRows()
+    // Evaluate GPU input expressions and get GPU columns
+    val gpuInputColumns = gpuInputs.safeMap(_.columnarEval(batch))
+
+    // Time the CPU processing (columnar->row->CPU expr->columnar)
+    val processingStartTime = System.nanoTime()
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala]: nit, we could use `GpuMetric.ns(metric) { }`, but it doesn't take an optional metric.

File: tests/src/test/scala/com/nvidia/spark/rapids/GpuCpuBridgeSuite.scala [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    )
+    
+    assert(bridgeExpr.prettyName == "gpu_cpu_bridge")
+  }
+  
+  // ============================================================================
+  // Configuration Tests
+  // ============================================================================
+  
+  test("Bridge config controls feature enablement") {
```
Comment (abellina) [inline tests/src/test/scala/com/nvidia/spark/rapids/GpuCpuBridgeSuite.scala]: this test is probably not necessary. What I think we need is testing the fallback, either in a suite or integration test.

File: tests/src/test/scala/org/apache/spark/sql/rapids/BridgeUnsafeProjectionSuite.scala [hunk 6]
```diff
@@ -0,0 +1,1373 @@
...
+  }
+
+  test("float type") {
+    val testData = Seq(
+      (1.5f, false),
+      (Float.MinValue, false),
+      (Float.MaxValue, false),
+      (0.0f, true), // null
+      (3.14159f, false),
+      (Float.NaN, false),
```
Comment (abellina) [inline tests/src/test/scala/org/apache/spark/sql/rapids/BridgeUnsafeProjectionSuite.scala:223]: should we also test negative floats and NaN ranges? I know we teted this in the past (SparkQueryCompareTestSuite) has float positive/negative nan lower/upper values.

File: tests/src/test/scala/org/apache/spark/sql/rapids/BridgeUnsafeProjectionSuite.scala [hunk 7]
```diff
@@ -0,0 +1,1373 @@
...
+  }
+
+  test("double type") {
+    val testData = Seq(
+      (1.5, false),
+      (Double.MinValue, false),
+      (Double.MaxValue, false),
+      (0.0, true), // null
+      (3.141592653589793, false),
+      (Double.NaN, false),
```
Comment (abellina) [inline tests/src/test/scala/org/apache/spark/sql/rapids/BridgeUnsafeProjectionSuite.scala:262]: same question about nan ranges and sign.

File: tests/src/test/scala/org/apache/spark/sql/rapids/BridgeUnsafeProjectionSuite.scala [hunk 8]
```diff
@@ -0,0 +1,1373 @@
...
+    )
+  }
+
+  test("timestamp type") {
+    val testData = Seq(
+      (1609459200000000L, false), // 2021-01-01 00:00:00 UTC
+      (0L, false),                 // 1970-01-01 00:00:00 UTC
+      (-1000000L, false),          // Before epoch
+      (0L, true),                  // null
+      (1672531200000000L, false)   // 2023-01-01 00:00:00 UTC
```
Comment (abellina) [inline tests/src/test/scala/org/apache/spark/sql/rapids/BridgeUnsafeProjectionSuite.scala]: how about h/m/s and ms?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala [hunk 9]
```diff
@@ -558,6 +603,8 @@ abstract class DataWritingCommandMeta[INPUT <: DataWritingCommand](
     if (checkTimeZone()) {
       timezoneCheck()
     }
+    // Update our expressions to allow them to run with the bridge if possible
+
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala]: nit extra space

File: tests/src/test/scala/org/apache/spark/sql/rapids/BridgeUnsafeProjectionSuite.scala [hunk 10]
```diff
@@ -0,0 +1,1373 @@
...
+        if (expectedIsNull) {
+          assert(result.isNull(idx))
+        } else {
+          assert(result.getLong(idx) == expectedValue)
+        }
+      }
+    )
+  }
+
+  test("multiple expressions") {
```
Comment (abellina) [inline tests/src/test/scala/org/apache/spark/sql/rapids/BridgeUnsafeProjectionSuite.scala:457]: should we add `DayTimeIntervalType` and `YearMonthIntervalType` and possibly an unsupported type?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala [hunk 11]
```diff
@@ -0,0 +1,940 @@
...
+  def createCode(
+      ctx: CodegenContext,
+      expressions: Seq[Expression],
+      useSubexprElimination: Boolean = false): ExprCode = {
+    val exprEvals = ctx.generateExpressions(expressions, useSubexprElimination)
+    val exprSchemas = expressions.map(e => Schema(e.dataType, e.nullable))
+
+    val numVarLenFields = exprSchemas.count {
+      case Schema(dt, _) => !UnsafeRow.isFixedLength(dt)
+      // TODO: consider large decimal and interval type
```
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala]: is there a follow on for this TODO?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala [hunk 12]
```diff
@@ -0,0 +1,940 @@
...
+             |  ${setNull.trim}
+             |} else {
+             |  ${writeField.trim}
+             |}
+           """.stripMargin
+        }
+    }
+
+    val writeFieldsCode = if (isTopLevel && (row == null || ctx.currentVars != null)) {
+      // TODO: support whole stage codegen
```
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala]: follow on for this TODO?

Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala]: Sorry the TODOs ended up in dead code before we went to inserting the data directly into the columnar data structures directly.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala (hunk 11 above)
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala]: Sorry the TODOs ended up in dead code before we went to inserting the data directly into the columnar data structures directly.

**Comment by revans2** [comment]
@abellina I think I have addressed all of your comments. Please take another look

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+ */
+case class GpuCpuBridgeExpression(
+    gpuInputs: Seq[Expression],
+    cpuExpression: Expression,
+    outputDataType: DataType,
+    outputNullable: Boolean) extends GpuExpression with ShimExpression 
+    with Logging with GpuBind with GpuMetricsInjectable {
+
+  // Only GPU inputs are children for GPU expression tree traversal
+  override def children: Seq[Expression] = gpuInputs ++ Seq(cpuExpression)
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala:53]: That is not true. A `GpuExpression` is an `Expression`. All Spark cares about is that children is a Seq of Expressions. We have to be vary careful that we always go through the bridge to be able to execute these CPU expressions properly, but that is a separate concern. This was a change we made recently to fix other bugs in how Spark planned things to make it simpler to actually traverse the entire tree properly.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala [hunk 14]
```diff
@@ -0,0 +1,648 @@
...
+      expressions: Seq[Expression],
+      subexpressionEliminationEnabled: Boolean): BridgeUnsafeProjection = {
+    // Check upfront if we can support all types with direct codegen
+    // This provides a fast-fail mechanism before attempting expensive codegen
+    val unsupportedTypes = expressions.map(_.dataType).filterNot(canSupportDirectCodegen)
+    if (unsupportedTypes.nonEmpty) {
+      throw new UnsupportedOperationException(
+        s"Cannot generate code for types: ${unsupportedTypes.mkString(", ")}. " +
+        "Falling back to interpreted evaluation.")
+    }
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala:570]: This is still a WIP, and will be fixed in a later PR.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+    }
+    
+    // Time the entire CPU bridge operation from the SparkPlan perspective
+    GpuMetric.nsOption(cpuBridgeWaitTime) {
+      // Currently sequential processing only
+      logDebug(s"Using sequential processing for ${numRows} rows " +
+        s"(expression: ${cpuExpression.getClass.getSimpleName})")
+      evaluateSequentially(batch)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala:144]: It does not matter in this case because 0 rows means that there will be NO rows. This is the equivalent of allocate empty, but it is just not as clean.

**Review by abellina — DISMISSED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+        NoopMetric,
+        NoopMetric,
+        nullSafe = false,
+        releaseSemaphore = false
+      )
+
+      // Delegate to evaluation function - ColumnarToRowIterator manages the batch lifecycle
+      evaluationFunction(rowIterator, numRows)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala:172]: That is not technically true, but I think the leak risk is okay. We have a set of input columns that we place inside of a ColumnarBatch, which is then placed inside of a single iterator, which is placed inside of a ColumnarToRowIterator. The ColumanrToRowIterator takes ownership of any batches it pulls out of the input iterator passed to it. So really we would only leak a batch if we got an exception in between creating the batch (line 155) and the first time next is called on the ColumanrToRowIterator, which is very quickly. I could be more exact in avoiding this possible leak, but it is so unlikely to happen, that I am okay with it.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+      val inputBatch = new ColumnarBatch(gpuInputColumns.toArray, numRows)
+      val rowIterator = new ColumnarToRowIterator(
+        Iterator.single(inputBatch),
+        NoopMetric,
+        NoopMetric,
+        NoopMetric,
+        NoopMetric,
+        nullSafe = false,
+        releaseSemaphore = false
+      )
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala:167]: This is by design. It is problematic right now and because of those problems the feature is off by default until all of the work is done. This is one of many PRs because the original PR was too large to review all at once. If we released the semaphore here, then we would have unspillable data held on the GPU, which means we would quickly run out of GPU memory as more tasks pile onto the GPU. Instead we take a performance hit and execute CPU code while holding the GPU semaphore. This is not ideal, but in a later PR will execute these expressions in a thread pool, so that we can finish the execution much faster, and at the same time it reduces total GPU memory pressure because we will not let other tasks onto the GPU at the same time. Yes it is not ideal because we are holding the semaphore while running CPU code, but from a systems standpoint it ends up being better overall.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala [hunk 18]
```diff
@@ -0,0 +1,648 @@
...
+  
+  override def apply(rows: Iterator[InternalRow],
+    builders: Array[RapidsHostColumnBuilder]): Unit = {
+    rows.foreach { row =>
+      expressions.zipWithIndex.foreach { case (expr, idx) =>
+        val result = expr.eval(row)
+        appendFunctions(idx)(result, builders(idx))
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala:62]: The builders are going to be thrown away at the end. Why do we need to properly clean them up when we are going to throw them away? all we care about is that the builders can release all of their resources without any errors in this case.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala:53]: We already covered this and you are wrong see https://github.com/NVIDIA/spark-rapids/pull/14003#discussion_r2677321168

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala [hunk 19]
```diff
@@ -1319,11 +1403,171 @@ abstract class BaseExprMeta[INPUT <: Expression](
...
+        }
+    }
+    val boundCpuExpression = expr.withNewChildren(boundChildren)
+
+    GpuCpuBridgeExpression(
+      gpuInputs = deduplicatedGpuInputs,
+      cpuExpression = boundCpuExpression,
+      outputDataType = expr.dataType,
+      outputNullable = expr.nullable)
+  }
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala:1585]: The code technically was correct, but this makes it clearer that it is correct and adds some tests to prove it.

**Comment by revans2** [comment]
@abellina sorry to bug you on this again. I added another test and cleaned up a little code to address a greptile comment.

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala:53]: Actually this looks like a real bug?

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+      // Handle empty batch case
+      return GpuColumnVector.fromNull(numRows, dataType)
+    }
+    
+    // Time the entire CPU bridge operation from the SparkPlan perspective
+    GpuMetric.nsOption(cpuBridgeWaitTime) {
+      // Check if we should use parallel processing
+      // Note: Nondeterministic expressions are excluded from CPU bridge entirely,
+      // so all expressions reaching this point are safe for parallel processing
+      if (GpuCpuBridgeThreadPool.shouldParallelize(numRows)) {
```
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala]: agree with greptile here. @revans2 did you push a WIP diff that we need to revert?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala [hunk 21]
```diff
@@ -0,0 +1,409 @@
+/*
+ * Copyright (c) 2026, NVIDIA CORPORATION.
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala:2]: Nope we are now in the year 2026

**Comment by revans2** [comment]
@abellina I somehow messed up an upmerge on another branch with this and got compile errors. That is now fixed. Please take another look

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/TypeConverter.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+   */
+  public abstract double append(SpecializedGetters row, int column, RapidsHostColumnBuilder builder);
+
+  /**
+   * Append row value to the column builder, but no value is returned as a performance
+   * optimization for scala.
+   */
+  public void appendNoRet(SpecializedGetters row, int column, RapidsHostColumnBuilder builder) {
+    append(row, column, builder);
+  }
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/TypeConverter.java:36]: Scala does auto-boxing and auto-unboxing of doubles. In many cases it is able to do a great job and there is no performance impact. In this specific case we saw a lot of performance problems from this, which is why this is in java, so that java can throw away the return type without the boxing/unboxing that scala would add if it were in there.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala [hunk 23]
```diff
@@ -1319,11 +1403,185 @@ abstract class BaseExprMeta[INPUT <: Expression](
...
+  /**
+   * Deduplicate GPU inputs using semantic equality to reduce memory transfers and code size.
+   * @param gpuInputsWithIndex GPU expressions paired with their original child indices
+   * @return (deduplicated GPU inputs, mapping from original indices to deduplicated indices)
+   */
+  private def deduplicateGpuInputs(
+      gpuInputsWithIndex: Seq[(Expression, Int)]): (Seq[Expression], Map[Int, Int]) = {
+    
+    import org.apache.spark.sql.rapids.catalyst.expressions.GpuExpressionEquals
+
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala:1428]: How? What are the cases when semanitc equality is not actual equality? If that is the case we have a lot of bugs in a lot of other locations in the code.

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala [hunk 24]
```diff
@@ -0,0 +1,648 @@
...
+             |} else {
+             |  $builderRef.append($valueVar.toJavaBigDecimal());
+             |}
+           """.stripMargin
+        } else {
+          s"$builderRef.append($valueVar.toJavaBigDecimal());"
+        }
+      
+      case NullType =>
+        s"$builderRef.appendNull();"
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala:343]: You are self conflicting in these comments. This one says that I need to add NullType to line 255, but line 255 has a null type handling in it. The comment about line 255 says this needs to add NullType handling, but it is here already.

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+    ThreadLocal.withInitial(() => {
+      // Register cleanup when this thread first accesses the ThreadLocal in this task
+      // Only register if we're actually in a Spark task (not in tests)
+      Option(org.apache.spark.TaskContext.get()).foreach { _ =>
+        onTaskCompletion {
+          threadLocalProjection.remove()
+        }
+      }
+      createCodeGeneratedProjection()
+    })
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCpuBridgeExpression.scala:125]: I disagree here. Each expression/task gets a separate `theadLocalProjection`. They are not shared between tasks, so the thread reuse is not an issue.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala [hunk 26]
```diff
@@ -0,0 +1,648 @@
...
+         |    }
+         |  }
+         |
+         |  ${ctx.declareAddedFunctions()}
+         |}
+       """.stripMargin
+
+    val code = CodeFormatter.stripOverlappingComments(
+      new CodeAndComment(codeBody, ctx.getPlaceHolderToComments()))
+    logDebug(s"code for ${expressions.mkString(",")}:\n${CodeFormatter.format(code)}")
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/BridgeGenerateUnsafeProjection.scala:643]: This already happens in so many other places and in so many other ways with Spark that I don't consider it a real security issue.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala [hunk 27]
```diff
@@ -1534,17 +1534,23 @@ abstract class BaseExprMeta[INPUT <: Expression](
   }
 
   def convertForGpuCpuBridge(): GpuExpression = {
-    if (!this.wrapped.deterministic) {
-      throw new IllegalStateException(s"$this is NOT deterministic!!!")
-    }
+    assert(this.wrapped.deterministic,
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala]: should this be a `require`?

Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala]: You are correct a scala assert does not always do way we want/expect.

**Comment by revans2** [comment]
build

**Review by abellina — APPROVED** [review]

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
build

**Comment by pxLi** [comment]
build
